File: src/common/common/main/java/settingdust/mod_sets/util/SavingData.java
package settingdust.mod_sets.util;

import com.google.common.base.Supplier;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

public interface SavingData {
    Gson GSON = new GsonBuilder()
        .setPrettyPrinting()
        .create();

    SavingData invoker = new SavingData() {
        private final Supplier<Iterable<SavingData>> services = () -> ServiceLoaderUtil.findServices(SavingData.class);

        @Override
        public void load() {
            for (final var data : services.get()) {
                data.load();
            }
        }

        @Override
        public void save() {
            for (final var data : services.get()) {
                data.save();
            }
        }
    };

    void load();

    void save();
}

File: src/common/main/kotlin/settingdust/mod_sets/data/SavingData.kt
package settingdust.mod_sets.data

import settingdust.mod_sets.util.ServiceLoaderUtil

interface SavingData {
    companion object : SavingData {
        private val instances by lazy { ServiceLoaderUtil.findServices<SavingData>() }

        override fun reload() {
            for (data in instances) {
                data.reload()
            }
        }

        override fun save() {
            for (data in instances) {
                data.save()
            }
        }
    }

    fun reload()

    fun save()
}
